source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1237 @@
+                      // Only apply sediment rain to areas below sea level.
+                      if (h[i] < sl)
+                        {
+                          // If the rain would put us above sea level, set height to sea level.
+                          if (h[i] + sediment_rain*a_dt > sl)
+                            h[i] = sl;
+                          else
+                            h[i] = h[i] + sediment_rain*a_dt;
+                        }
+                    }
+                }
+
+
+              /*
+               * The ghost nodes are added as a single layer of nodes surrounding the entire model,
+               * so that 3x3 amount of nodes representing ASPECT will become 5x5 in FastScape, with the inner
+               * 3x3 being returned to ASPECT.
+               */
+              // I redid the indexing here, at some point I should double check that these still work without issue.
+              if (use_ghost)
+                 set_ghost_nodes(h.get(), vx.get(), vy.get(), vz.get(), nx, ny);
+
+
+              // Get current fastscape timestep.
+              int istep = 0;
+              fastscape_get_step_(&istep);
+
+              // Write a file to store h & step in case of restart.
+              // TODO: there's probably a faster way to write these.
+              if ((this->get_parameters().checkpoint_time_secs == 0) &&
+                  (this->get_parameters().checkpoint_steps > 0) &&
+                  (current_timestep % this->get_parameters().checkpoint_steps == 0))
+                {
+
+                  std::ofstream out_h (restart_filename.c_str());
+                  std::ofstream out_step (restart_step_filename.c_str());
+                  std::ofstream out_b (restart_filename_basement.c_str());
+                  std::stringstream bufferb;
+                  std::stringstream bufferh;
+
+                  fastscape_copy_basement_(b.get());
+
+                  out_step << (istep+restart_step) << "\n";
+
+                  for (int i=0; i<array_size; i++)
+                    {
+                      bufferh << h[i] << "\n";
+                      bufferb << b[i] << "\n";
+                    }
+
+                  out_h << bufferh.str();
+                  out_b << bufferb.str();
+                }
+
+              // Find a fastscape timestep that is below our maximum timestep.
+              int steps = nstep;
+              double f_dt = a_dt/steps;
+              while (f_dt>max_timestep)
+                {
+                  steps=steps*2;
+                  f_dt = a_dt/steps;
+                }
+
+              // Set time step
+              fastscape_set_dt_(&f_dt);
+
+              // Set velocity components and h.
+              if(use_v)
+              {
+                fastscape_set_u_(vz.get());
+                fastscape_set_v_(vx.get(), vy.get());
+              }
+
+              fastscape_set_h_(h.get());
+
+              // The visualization is 1 step behind the current timestep.
```

Comment:
I set it this way because on the first timestep, we create an initial VTK file before running FastScape and a second one after running it. Though, the visualization step only changes the file name on the VTK files.